### ISSUE TITLE
Support for others data sample format

### ISSUE THREAD

**Issue by brunodoamaral** [Statoil/segyio#238]
Hi,

First of all, thanks for sharing this lib. It is a very nice tool for reading seg-y files.

I'd like to help in getting the lib to work with other data sample formats. I was planning to use a public survey to show some seg-y features. I choose to use the [Netherlands Offshore F3 Block - Complete](https://www.opendtect.org/osr/Main/NetherlandsOffshoreF3BlockComplete4GB) data, but once I tried to open its raw seg-y, it gaves me the error `RuntimeError: trace count inconsistent with file size, trace lengths possibly of non-uniform`.

Digging into the issue, I discovered the snippet bellow:

https://github.com/Statoil/segyio/blob/91c46193ffabf586773d1d8125888dbc20c19985/lib/src/segy.c#L712-L716

It is assumed that all trace sample is 4-bytes, but the one I was using is 2-bytes.

I understand that this function should be updated to receive the format, as well as the number of samples, in order to correctly calculate the `trace_bsize`.

I think that this is the 'easy' part, as there aren't a lot of references to this function on the code. However, it seems that treating each format in code should be the hardest part. As this one:

https://github.com/Statoil/segyio/blob/4f1fcfdaaa25d6e608b1d47e2f175f9a81f52cdc/applications/segyinfo.c#L102-L140

Here, one assumes that `trbuf` is a float number and calculate the maximum of trace value. If, instead, the trace is a 2-byte int, than the result will be noise.

I opened this ticket to start a discussion regarding the possible solutions to the issue *and* if the maintainers think that this fix is worth.  

Best regards!
Bruno

**Comment by jokva** [comment]
Hi,

It's certainly possible to support other stuff than 4-byte floats, but I'm not sure if we'll be able to do it without breaking interfaces. Not accepting a variable-width format is one of the bigger regrets from the early design phase, it should've been in the interface from the get-go.

I'll have a quick look over the source code and see if I can figure out exactly what it takes to be able to nicely support non-4-byte inputs.

**Comment by jokva** [comment]
From a *very* quick glance it doesn't seem to be that difficult to support more formats, although it might be slightly inelegant to do it in a way that doesn't break any interfaces. I'll look a bit more into it.

**Comment by brunodoamaral** [comment]
Hi @jokva! I spend some time today in order to make a proof-of-concept implementation of multiple sample format.
I forked your project and commited the changes here: https://github.com/brunodoamaral/segyio/commit/dd026ecb6de38caf48dd41f4d15c0c978b6cf32c

For now, it just works for `segyinfo` application. I was able to successfully run this code on both float32 and int16 segy files.

In a nutshell, I added a new parameter to some `segyio.c` functions (`int sample_bsize`) and changed pointers from `float*` to `void*`. It helped the compiler alert me of flows in accessing the data.

I think that my proposed core implementation is somehow reasonable, but the changes I have to made on `segyinfo.c` were definitely *not elegant*.

Let me know your opinion on the changes I made and, if you find it useful, I can continue adding support for more applications and the Python interface.

**Comment by brunodoamaral** [comment]
Ok, this got my attention: I could access all traces in a seg-y file using pure numpy.
There are a few gotchas, but it works for int16 and float32 files (in IEEE format - for IBM float we need to [convert](https://mail.scipy.org/pipermail/scipy-user/2009-January/019392.html)).
I just want to share this piece of code because for my usage, which is pure read-only, if works perfectly

```python
import numpy as np

segy_file = '../input/F3_Demo_2016_training_v6/Rawdata/Seismic_data.sgy'

# Some definitions
SEGY_BINARY_HEADER_SIZE = 400
SEGY_TEXT_HEADER_SIZE = 3200
SEGY_TRACE_HEADER_SIZE = 240

def np_traces(fname, trace0, n_samples, sample_format):
    segy_trace_dtype = np.dtype([
        ('header', (bytes, SEGY_TRACE_HEADER_SIZE)),
        ('trace',  (sample_format, n_samples))
    ])
    
    segy_raw = np.memmap(segy_file, dtype=segy_trace_dtype, offset=trace0, mode='r')
    
    return segy_raw

trace0 = SEGY_BINARY_HEADER_SIZE+SEGY_TEXT_HEADER_SIZE
n_samples = 462
sample_format = '>i2'

segy_raw = np_traces(segy_file, trace0, n_samples, sample_format)
all_traces = segy_raw['trace']
a_trace = all_traces[10000]

print(segy_raw.shape)   # outputs (600515,)
print(all_traces.shape) # outputs (600515, 462)
print(a_trace.shape)    # outputs (462,)
```

Performance-wise, it seems to be very fast:
```python
%time segy_raw['trace'].max()
# CPU times: user 456 ms, sys: 70.3 ms, total: 526 ms
# Wall time: 523 ms
# 32767
```

```python
%time segy_raw['trace'].min()
# CPU times: user 442 ms, sys: 3.94 ms, total: 446 ms
# Wall time: 443 ms
# -32767
```

**Comment by jokva** [comment]
Hi, Bruno.

The numpy trick is very cool (we have a few people using a similar technique), and it works well if all you care about is reading some data. segyio obviously tries to accomplish a lot more than that, but it's a cool technique and a useful debugging tool (I've used it myself, actually).

**Comment by jokva** [comment]
I'm going to go through your proposed changes sometime later, but from a glance your style seems fine.

I'm not comfortable changing the signature from `float*` to `void*`, however, and it makes some interop code a lot worse to work with, but I have a few ideas on how we can get support for the rest of the formats, without breaking the `float*` interface.

It will require a new `trace_bsize` style function, at minimum.

**Comment by jokva** [comment]
Hi Bruno, how's this work going?

I *think* we'll be able to update all read/write signatures to work on `void*` instead of `float*` without breaking API and ABI (and at least on gcc, it doesn't create much noise with implicit-void casts either). It obviously means taking some care rewriting iteration and especially conversion functions, but all-in-all I think it's feasible.

**Comment by brunodoamaral** [comment]
Hi @jokva! I had a deliverable last week that used the *Netherlands Offshore F3 Block* data. As it was a read-only code, I used the numpy-version to read the file.

Regarding segyio, I just get in touch with Python bind code. My change in API made "only" 5 compile errors - but more bugs may be hidden in Python code.
I'll fix them and push changes to my repository.

In the meantime, it would be great to have a test-case written in Python and/or C that does some work on non-float32 files. I'd suggest the following test that run for every supported trace format (int8, int16, int32, float32, float32-ibm):

1. Create some random traces in the chosen format;
1. Write to a temporary segy file;
1. Read it back to memory and compare with original random traces;
1. Read segy file using the numpy-trick and compare with original traces;
1. Remove the temporary file.

**Comment by brunodoamaral** [comment]
Regarding the `void *` usage, it has the advantage of generating **useful** compiler warnings: iterating over `void *` gives the warning that its size is not defined. Correctly resolving these issues are one important step.

**Comment by jokva** [comment]
> Regarding segyio, I just get in touch with Python bind code. My change in API made "only" 5 compile errors - but more bugs may be hidden in Python code.

Maybe. The change itself, save some pointer manipulation stuff, isn't too large and shouldn't affect much, but the difference is really gonna show in the `native` family of functions. I'm pretty sure I know how to do that portably and standard compliant, but it's an exercise for sure.

> In the meantime, it would be great to have a test-case written in Python and/or C that does some work on non-float32 files.

100% behind you. I think an idea is to do that, and create a similarly predictable file as the bundled `small.sgy` and run most of the tests on different numerical types.

**Comment by brunodoamaral** [comment]
Hi @jokva. I just pushed the modifications that *worked on my machine* here: https://github.com/brunodoamaral/segyio/tree/multiple-sample-data-format
It passes all current tests.
Next step is to create the test code described [before](https://github.com/Statoil/segyio/issues/238#issuecomment-374212534). But for it, I need to be more familiar with creating the segy file.
I'm planning to do it along this week. But feel free to make some tests with the branch I created.

**Comment by jokva** [comment]
Hi Bruno,

Inspired by your branch I started experimenting with the core library and support for more formats. I opened #241 to showcase it and gather some feedback.

At the time of writing, all changes are in the core library, and the python extension is untouched. It preserves API and ABI compatibility, but in order to properly use the new formats, two new functions are added.

What's really lacking currently are some tests, both files and implementations. I couldn't find anything in your branch, but your numpy snippet will be very useful when constructing it.

Thanks for your help so far, I hope you'll be able to help both develop and test this new feature with me.
